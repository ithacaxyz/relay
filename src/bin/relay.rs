//! # Ithaca Relay
//!
//! A relay service that sponsors transactions for EIP-7702 accounts.

use clap::Parser;
use relay::{
    cli::*,
    otlp::{OtelConfig, OtelGuard},
};
use tracing::debug;
use tracing_subscriber::prelude::*;

fn init_tracing_subscriber() -> Option<OtelGuard> {
    let registry = tracing_subscriber::registry().with(
        tracing_subscriber::fmt::layer()
            .with_filter(tracing_subscriber::filter::EnvFilter::from_default_env()),
    );

    if let Some(cfg) = OtelConfig::load() {
        let guard = cfg.provider();
        registry.with(guard.layer()).init();
        Some(guard)
    } else {
        registry.init();
        None
    }
}

#[tokio::main]
async fn main() {
    // Enable backtraces unless a RUST_BACKTRACE value has already been explicitly provided.
    if std::env::var_os("RUST_BACKTRACE").is_none() {
        unsafe { std::env::set_var("RUST_BACKTRACE", "1") };
    }

    let _guard = init_tracing_subscriber();
    if _guard.is_some() {
        debug!("opentelemetry initialized");
    }

    let args = Args::parse();
    if let Err(err) = args.run().await {
        eprintln!("Error: {err:?}");
        std::process::exit(1);
    }
}
