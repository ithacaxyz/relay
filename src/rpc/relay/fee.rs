use crate::{
    chains::Chain,
    error::{KeysError, QuoteError, RelayError},
    provider::ProviderExt,
    rpc::relay::manager::Manager,
    types::{
        Account, AssetDiffs, GasEstimate, Intent, IntentKind, Key, KeyWith712Signer, Orchestrator,
        OrchestratorContract, PartialAction, Quote, Signature, SignedCalls, Transfer,
    },
};
use alloy::{
    consensus::{SignableTransaction, TxEip1559},
    eips::eip7702::constants::EIP7702_DELEGATION_DESIGNATOR,
    primitives::{Address, Bytes, U256, bytes},
    providers::{
        Provider,
        utils::{EIP1559_FEE_ESTIMATION_PAST_BLOCKS, Eip1559Estimator},
    },
    rpc::types::state::{AccountOverride, StateOverridesBuilder},
    sol_types::{SolCall, SolValue},
};
use futures_util::{TryFutureExt, future::try_join4};
use tracing::{debug, instrument};

#[async_trait::async_trait]
pub trait FeeEstimator: Manager {
    /// Estimate fees for executing actions
    #[instrument(skip_all)]
    async fn estimate_fee(
        &self,
        request: PartialAction,
        token: Address,
        authorization_address: Option<Address>,
        account_key: Key,
        key_slot_override: bool,
        intent_kind: IntentKind,
    ) -> Result<(AssetDiffs, Quote), RelayError> {
        let chain = self
            .chains()
            .get(request.chain_id)
            .ok_or(RelayError::UnsupportedChain(request.chain_id))?;

        let provider = chain.provider.clone();
        let Some(token) = self.fee_tokens().find(request.chain_id, &token) else {
            return Err(QuoteError::UnsupportedFeeToken(token).into());
        };

        // create key
        let mock_key = KeyWith712Signer::random_admin(account_key.keyType)
            .map_err(RelayError::from)
            .and_then(|k| k.ok_or_else(|| RelayError::Keys(KeysError::UnsupportedKeyType)))?;

        // mocking key storage for the eoa, and the balance for the mock signer
        let overrides = StateOverridesBuilder::with_capacity(2)
            // simulateV1Logs requires it, so the function can only be called under a testing
            // environment
            .append(self.simulator(), AccountOverride::default().with_balance(U256::MAX))
            .append(self.orchestrator(), AccountOverride::default().with_balance(U256::MAX))
            .append(
                request.intent.eoa,
                AccountOverride::default()
                    .with_balance(U256::MAX.div_ceil(2.try_into().unwrap()))
                    .with_state_diff(if key_slot_override {
                        account_key.storage_slots()
                    } else {
                        Default::default()
                    })
                    // we manually etch the 7702 designator since we do not have a signed auth item
                    .with_code_opt(authorization_address.map(|addr| {
                        Bytes::from([&EIP7702_DELEGATION_DESIGNATOR, addr.as_slice()].concat())
                    })),
            )
            .build();

        let account = Account::new(request.intent.eoa, &provider).with_overrides(overrides.clone());

        let (orchestrator, delegation, fee_history, eth_price) = try_join4(
            // fetch orchestrator from the account and ensure it is supported
            async {
                let orchestrator = account.get_orchestrator().await?;
                if !self.is_supported_orchestrator(&orchestrator) {
                    return Err(RelayError::UnsupportedOrchestrator(orchestrator));
                }
                Ok(Orchestrator::new(orchestrator, &provider).with_overrides(overrides.clone()))
            },
            // fetch delegation from the account and ensure it is supported
            self.has_supported_delegation(&account).map_err(RelayError::from),
            // fetch chain fees
            provider
                .get_fee_history(
                    EIP1559_FEE_ESTIMATION_PAST_BLOCKS,
                    Default::default(),
                    &[self.priority_fee_percentile()],
                )
                .map_err(RelayError::from),
            // fetch price in eth
            async {
                // TODO: only handles eth as native fee token
                Ok(self.price_oracle().eth_price(token.kind).await)
            },
        )
        .await?;

        let native_fee_estimate = Eip1559Estimator::default().estimate(
            fee_history.latest_block_base_fee().unwrap_or_default(),
            &fee_history.reward.unwrap_or_default(),
        );

        let Some(eth_price) = eth_price else {
            return Err(QuoteError::UnavailablePrice(token.address).into());
        };
        let payment_per_gas = if intent_kind.is_single() {
            (native_fee_estimate.max_fee_per_gas as f64 * 10u128.pow(token.decimals as u32) as f64)
                / f64::from(eth_price)
        } else {
            // todo: is_multi_input should take a fee as well eventually
            0f64
        };

        // fill intent
        let mut intent = Intent {
            eoa: request.intent.eoa,
            executionData: request.intent.execution_data.clone(),
            nonce: request.intent.nonce,
            payer: request.intent.payer.unwrap_or_default(),
            paymentToken: token.address,
            paymentRecipient: self.fee_recipient(),
            supportedAccountImplementation: delegation,
            encodedPreCalls: request
                .intent
                .pre_calls
                .into_iter()
                .map(|pre_call| pre_call.abi_encode().into())
                .collect(),
            encodedFundTransfers: request
                .intent
                .fund_transfers
                .into_iter()
                .map(|(token, amount)| Transfer { token, amount }.abi_encode().into())
                .collect(),
            ..Default::default()
        };

        let extra_payment = self.estimate_extra_fee(&chain, &intent).await?
            * U256::from(10u128.pow(token.decimals as u32))
            / eth_price;

        let intrinsic_gas = approx_intrinsic_cost(
            &OrchestratorContract::executeCall {
                isMultiChain: !intent_kind.is_single(),
                encodedIntent: intent.abi_encode().into(),
            }
            .abi_encode(),
            authorization_address.is_some(),
        );

        let initial_payment = U256::from(intrinsic_gas as f64 * payment_per_gas) + extra_payment;

        intent.set_legacy_payment_amount(initial_payment);

        // sign intent
        let signature = mock_key
            .sign_typed_data(
                &intent.as_eip712().map_err(RelayError::from)?,
                &orchestrator
                    .eip712_domain(intent.is_multichain())
                    .await
                    .map_err(RelayError::from)?,
            )
            .await
            .map_err(RelayError::from)?;

        intent.signature = Signature {
            innerSignature: signature,
            keyHash: account_key.key_hash(),
            prehash: request.prehash,
        }
        .abi_encode_packed()
        .into();

        if !intent.encodedFundTransfers.is_empty() {
            // todo: the contract version is broken, and any signature will pass.
            intent.funder = self.contracts().funder.address;
        }

        // todo: simulate with executeMultiChain if intent.is_multichain
        // we estimate gas and fees
        let (asset_diff, sim_result) = orchestrator
            .simulate_execute(
                self.simulator(),
                &intent,
                account_key.keyType,
                payment_per_gas,
                self.asset_info().clone(),
            )
            .await?;

        // todo: re-evaluate if this is still necessary
        let gas_estimate = GasEstimate::from_combined_gas(
            sim_result.gCombined.to(),
            intrinsic_gas,
            self.quote_config(),
        );

        debug!(eoa = %request.intent.eoa, gas_estimate = ?gas_estimate, "Estimated intent");

        // Fill combinedGas and empty dummy signature
        intent.combinedGas = U256::from(gas_estimate.intent);
        intent.signature = bytes!("");
        intent.funderSignature = bytes!("");

        // Calculate amount with updated paymentPerGas
        if !intent_kind.is_single() {
            // todo: temporary
            intent.set_legacy_payment_amount(U256::ZERO);
        } else {
            intent.set_legacy_payment_amount(
                extra_payment + U256::from((payment_per_gas * gas_estimate.tx as f64).ceil()),
            )
        }

        let quote = Quote {
            chain_id: request.chain_id,
            payment_token_decimals: token.decimals,
            output: intent,
            extra_payment,
            eth_price,
            tx_gas: gas_estimate.tx,
            native_fee_estimate,
            authorization_address,
            orchestrator: *orchestrator.address(),
            is_multi_chain: !intent_kind.is_single(),
        };

        Ok((asset_diff, quote))
    }

    /// Estimates additional fees to be paid for a intent (e.g L1 DA fees).
    ///
    /// Returns fees in ETH.
    #[instrument(skip_all)]
    async fn estimate_extra_fee(&self, chain: &Chain, intent: &Intent) -> Result<U256, RelayError> {
        // Include the L1 DA fees if we're on an OP rollup.
        let fee = if chain.is_optimism {
            // Create a dummy transactions with all fields set to max values to make sure that
            // calldata is largest possible
            let tx = TxEip1559 {
                chain_id: chain.chain_id,
                nonce: u64::MAX,
                gas_limit: u64::MAX,
                max_fee_per_gas: u128::MAX,
                max_priority_fee_per_gas: u128::MAX,
                to: (!Address::ZERO).into(),
                input: intent.encode_execute(false),
                ..Default::default()
            };
            let signature = alloy::signers::Signature::new(U256::MAX, U256::MAX, true);

            let encoded = {
                let tx = tx.into_signed(signature);
                let mut buf = Vec::with_capacity(tx.eip2718_encoded_length());
                tx.eip2718_encode(&mut buf);
                buf
            };

            chain.provider.estimate_l1_fee(encoded.into()).await?
        } else {
            U256::ZERO
        };

        Ok(fee)
    }
}

/// Approximates the intrinsic cost of a transaction.
///
/// This function assumes Prague rules.
fn approx_intrinsic_cost(input: &[u8], has_auth: bool) -> u64 {
    use alloy::eips::eip7702::constants::PER_EMPTY_ACCOUNT_COST;

    let zero_data_len = input.iter().filter(|v| **v == 0).count() as u64;
    let non_zero_data_len = input.len() as u64 - zero_data_len;
    let non_zero_data_multiplier = 4; // as defined in istanbul
    let standard_token_cost = 4;
    let tokens = zero_data_len + non_zero_data_len * non_zero_data_multiplier;

    // for 7702 designations there is an additional gas charge
    //
    // note: this is not entirely accurate, as there is also a gas refund in 7702, but at this
    // point it is not possible to compute the gas refund, so it is an overestimate, as we also
    // need to charge for the account being presumed empty.
    let auth_cost = if has_auth { PER_EMPTY_ACCOUNT_COST } else { 0 };

    21000 + auth_cost + tokens * standard_token_cost
}
