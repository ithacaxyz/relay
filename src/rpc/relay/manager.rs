use crate::{
    asset::AssetInfoServiceHandle,
    chains::Chains,
    config::QuoteConfig,
    error::{AuthError, RelayError},
    price::PriceOracle,
    storage::StorageApi,
    types::{Account, FeeTokens, VersionedContracts},
};
use alloy::{
    eips::eip7702::constants::EIP7702_DELEGATION_DESIGNATOR,
    primitives::{Address, Bytes, ChainId},
    providers::{DynProvider, Provider},
    rpc::types::state::StateOverridesBuilder,
};
use tracing::instrument;

#[async_trait::async_trait]
pub trait Manager {
    /// Get chains
    fn chains(&self) -> &Chains;

    /// Get contracts
    fn contracts(&self) -> &VersionedContracts;

    /// Get orchestrator address
    fn orchestrator(&self) -> Address {
        self.contracts().orchestrator.address
    }

    /// Previously deployed orchestrators.
    fn legacy_orchestrators(&self) -> impl Iterator<Item = Address> {
        self.contracts().legacy_orchestrators.iter().map(|c| c.address)
    }

    /// Get delegation implementation address
    fn delegation_implementation(&self) -> Address {
        self.contracts().delegation_implementation.address
    }

    /// Previously deployed delegation implementations.
    fn legacy_delegations(&self) -> impl Iterator<Item = Address> {
        self.contracts().legacy_delegations.iter().map(|c| c.address)
    }

    /// The simulator address.
    fn simulator(&self) -> Address {
        self.contracts().simulator.address
    }

    /// Get fee recipient
    fn fee_recipient(&self) -> Address;

    /// Get priority fee percentile
    fn priority_fee_percentile(&self) -> f64;

    /// Get quote config
    fn quote_config(&self) -> &QuoteConfig;

    /// Get fee tokens
    fn fee_tokens(&self) -> &FeeTokens;

    /// Get price oracle
    fn price_oracle(&self) -> &PriceOracle;

    /// Get asset info
    fn asset_info(&self) -> &AssetInfoServiceHandle;

    /// Get storage
    fn storage(&self) -> &dyn StorageApi;

    /// Get a provider for the given chain ID
    fn provider(&self, chain_id: ChainId) -> Result<DynProvider, RelayError> {
        self.chains()
            .get(chain_id)
            .map(|chain| chain.provider.clone())
            .ok_or(RelayError::UnsupportedChain(chain_id))
    }

    /// Checks if the orchestrator is supported.
    fn is_supported_orchestrator(&self, orchestrator: &Address) -> bool {
        self.orchestrator() == *orchestrator
            || self.legacy_orchestrators().any(|c| c == *orchestrator)
    }

    /// Checks if the account has a supported delegation implementation. If so, returns it.
    async fn has_supported_delegation<P: Provider + Clone>(
        &self,
        account: &Account<P>,
    ) -> Result<Address, RelayError> {
        let address = self.get_delegation_implementation(account).await?;
        if self.delegation_implementation() == address
            || self.legacy_delegations().any(|c| c == address)
        {
            return Ok(address);
        }
        Err(AuthError::InvalidDelegation(address).into())
    }

    /// Ensures the account has the latest delegation implementation. Otherwise, returns error.
    async fn ensure_latest_delegation<P: Provider + Clone>(
        &self,
        account: &Account<P>,
    ) -> Result<(), RelayError> {
        let address = self.has_supported_delegation(account).await?;
        if self.delegation_implementation() != address {
            return Err(AuthError::InvalidDelegation(address).into());
        }
        Ok(())
    }

    /// Returns the delegation implementation address from the requested account.
    ///
    /// It will return error if the delegation proxy is invalid.
    #[instrument(skip_all)]
    async fn get_delegation_implementation<P: Provider + Clone>(
        &self,
        account: &Account<P>,
    ) -> Result<Address, RelayError> {
        if let Some(delegation) = account.delegation_implementation().await? {
            return Ok(delegation);
        }

        // Attempt to retrieve the delegation proxy from storage, since it might not be
        // deployed yet.
        let Some(stored) = self.storage().read_account(&account.address()).await? else {
            return Err(RelayError::Auth(AuthError::EoaNotDelegated(account.address()).boxed()));
        };

        let address = account.address();
        let account = account.clone().with_overrides(
            StateOverridesBuilder::default()
                .with_code(
                    address,
                    Bytes::from(
                        [
                            &EIP7702_DELEGATION_DESIGNATOR,
                            stored.signed_authorization.address().as_slice(),
                        ]
                        .concat(),
                    ),
                )
                .build(),
        );

        account.delegation_implementation().await?.ok_or_else(|| {
            RelayError::Auth(
                AuthError::InvalidDelegationProxy(*stored.signed_authorization.address()).boxed(),
            )
        })
    }
}
