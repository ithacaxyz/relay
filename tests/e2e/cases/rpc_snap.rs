use crate::e2e::{
    AuthKind, ExpectedOutcome, MockErc20, TxContext, await_calls_status,
    cases::{upgrade_account_eagerly, upgrade_account_lazily},
    environment::{Environment, EnvironmentConfig},
};
use alloy::{
    primitives::{Address, B256, Bytes, U64, U256, b256},
    providers::Provider,
    sol_types::SolCall,
    uint,
};
use relay::{
    rpc::RelayApiClient,
    signers::Eip712PayLoadSigner,
    types::{
        Call, CallPermission, IERC20, KeyType, KeyWith712Signer,
        rpc::{
            AddFaucetFundsParameters, BundleId, GetAssetsParameters, GetAuthorizationParameters,
            GetKeysParameters, Meta, Permission, PrepareCallsCapabilities, PrepareCallsParameters,
            PrepareUpgradeAccountParameters, RequiredAsset, SendPreparedCallsParameters,
            UpgradeAccountCapabilities, UpgradeAccountParameters, UpgradeAccountSignatures,
            VerifySignatureParameters,
        },
    },
};
use std::str::FromStr;

const ADMIN_KEY: B256 = b256!("0x013c83ce5b08455e0505392b45f6d5effc7671e6f69108268ee4fedae1df72c8");
const SESSION_KEY: B256 =
    b256!("0xe1d7a91c93712db419ab647152bc1ae8739aca2f1bc8a6220d47fcf4a541a73c");

#[tokio::test]
async fn test_health() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let response = env.relay_endpoint.health().await?;
    insta::assert_json_snapshot!(response, {
        ".status" => reduction_from_str::<String>("status"),
        ".version" => reduction_from_str::<String>("version"),
    });

    Ok(())
}

#[tokio::test]
async fn test_get_capabilities() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let mut response =
        env.relay_endpoint.get_capabilities(Some(vec![U64::from(env.chain_id())])).await?;

    for (_, caps) in response.0.iter_mut() {
        caps.contracts.legacy_orchestrators.sort();
        caps.contracts.legacy_delegations.sort();
        caps.fees.tokens.sort_by_key(|token| token.uid.clone());
    }

    insta::assert_json_snapshot!(response);

    Ok(())
}

#[tokio::test]
async fn test_get_keys() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let admin_key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();
    let session_key = KeyWith712Signer::mock_session_with_key(KeyType::P256, SESSION_KEY)?
        .unwrap()
        .with_permissions(vec![Permission::Call(CallPermission {
            to: env.erc20,
            selector: MockErc20::transferCall::SELECTOR.into(),
        })]);

    upgrade_account_eagerly(&env, &[admin_key.to_authorized()], &admin_key, AuthKind::Auth).await?;

    // Add session key
    TxContext {
        authorization_keys: vec![&session_key],
        expected: ExpectedOutcome::Pass,
        key: Some(&admin_key),
        ..Default::default()
    }
    .process(0, &env)
    .await?;

    let response = env
        .relay_endpoint
        .get_keys(GetKeysParameters { address: env.eoa.address(), chain_ids: vec![] })
        .await?;
    insta::assert_json_snapshot!(response);

    Ok(())
}

#[tokio::test]
async fn test_get_assets() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let mut response = env
        .relay_endpoint
        .get_assets(GetAssetsParameters { account: env.eoa.address(), ..Default::default() })
        .await?;

    for (_, assets) in response.0.iter_mut() {
        assets.sort_by_key(|asset| asset.address);
    }

    insta::assert_json_snapshot!(response);

    Ok(())
}

#[tokio::test]
async fn test_prepare_calls() -> eyre::Result<()> {
    let config =
        EnvironmentConfig { num_chains: 2, fee_recipient: Address::ZERO, ..Default::default() };
    let env = Environment::setup_with_config(config.clone()).await?;

    // Create a key for signing
    let key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    // Account upgrade deployed onchain.
    upgrade_account_lazily(&env, &[key.to_authorized()], AuthKind::Auth).await?;

    let balance =
        IERC20::new(env.erc20, env.provider_for(1)).balanceOf(env.eoa.address()).call().await?
            / uint!(2_U256);
    let response = env
        .relay_endpoint
        .prepare_calls(PrepareCallsParameters {
            calls: vec![Call::transfer(env.erc20, Address::ZERO, uint!(1_U256))],
            chain_id: env.chain_id_for(0),
            from: Some(env.eoa.address()),
            capabilities: PrepareCallsCapabilities {
                authorize_keys: Default::default(),
                meta: Meta { fee_token: env.erc20, fee_payer: None, nonce: Some(U256::ZERO) },
                pre_calls: Default::default(),
                pre_call: Default::default(),
                required_funds: vec![RequiredAsset::new(env.erc20, balance)],
                revoke_keys: Default::default(),
            },
            balance_overrides: Default::default(),
            state_overrides: Default::default(),
            key: Some(key.to_call_key()),
        })
        .await?;

    let mut value = serde_json::to_value(response)?;
    sort_json_keys(&mut value);
    // TODO: fee related redactions can be removed if we make the tip block of Anvil at the moment
    // of this assertion consistent across runs, right now it's either 15 or 16, and this changes
    // the EIP-1559 fee estimations.
    insta::assert_json_snapshot!(value, {
        ".capabilities.assetDiffs.*" => insta::sorted_redaction(),
        ".capabilities.feeTotals.*.value" => reduction_from_str::<f64>("value"),
        ".context.quote.hash" => reduction_from_str::<B256>("hash"),
        ".context.quote.quotes[].intent.encodedPreCalls[]" => reduction_from_str::<Bytes>("encodedPreCall"),
        ".context.quote.quotes[].intent.prePaymentAmount" => reduction_from_str::<U256>("prePaymentAmount"),
        ".context.quote.quotes[].intent.prePaymentMaxAmount" => reduction_from_str::<U256>("prePaymentMaxAmount"),
        ".context.quote.quotes[].intent.totalPaymentAmount" => reduction_from_str::<U256>("totalPaymentAmount"),
        ".context.quote.quotes[].intent.totalPaymentMaxAmount" => reduction_from_str::<U256>("totalPaymentMaxAmount"),
        ".context.quote.quotes[].nativeFeeEstimate.maxFeePerGas" => reduction_from_str::<U256>("maxFeePerGas"),
        ".context.quote.r" => reduction_from_str::<U256>("r"),
        ".context.quote.s" => reduction_from_str::<U256>("s"),
        ".context.quote.ttl" => insta::dynamic_redaction(|value, _path| {
            assert!(value.as_u64().unwrap() > std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs());
            "[ttl]"
        }),
        ".context.quote.v" => reduction_from_str::<U64>("v"),
        ".context.quote.yParity" => reduction_from_str::<U64>("yParity"),
        ".digest" => reduction_from_str::<B256>("digest"),
        ".typedData.message.encodedPreCalls[]" => reduction_from_str::<Bytes>("encodedPreCall"),
        ".typedData.message.prePaymentMaxAmount" => reduction_from_str::<U256>("prePaymentMaxAmount"),
        ".typedData.message.totalPaymentMaxAmount" => reduction_from_str::<U256>("totalPaymentMaxAmount"),
    });

    Ok(())
}

#[tokio::test]
async fn test_prepare_upgrade_account() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let admin_key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    let response = env
        .relay_endpoint
        .prepare_upgrade_account(PrepareUpgradeAccountParameters {
            address: env.eoa.address(),
            delegation: env.delegation,
            chain_id: None,
            capabilities: UpgradeAccountCapabilities {
                authorize_keys: vec![admin_key.to_authorized()],
            },
        })
        .await?;

    // Nonces are random, so we need to redact them and the digest that depends on them.
    insta::assert_json_snapshot!(response, {
        ".digests.exec" => reduction_from_str::<B256>("digest"),
        ".context.preCall.nonce" => reduction_from_str::<B256>("nonce"),
        ".typedData.message.nonce" => reduction_from_str::<B256>("nonce"),
    });

    Ok(())
}

#[tokio::test]
async fn test_send_prepared_calls() -> eyre::Result<()> {
    let config =
        EnvironmentConfig { num_chains: 2, fee_recipient: Address::ZERO, ..Default::default() };
    let env = Environment::setup_with_config(config.clone()).await?;

    // Create a key for signing
    let key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    // Account upgrade deployed onchain.
    upgrade_account_lazily(&env, &[key.to_authorized()], AuthKind::Auth).await?;

    let balance =
        IERC20::new(env.erc20, env.provider_for(1)).balanceOf(env.eoa.address()).call().await?
            / uint!(2_U256);
    let response = env
        .relay_endpoint
        .prepare_calls(PrepareCallsParameters {
            calls: vec![Call::transfer(env.erc20, Address::ZERO, uint!(1_U256))],
            chain_id: env.chain_id_for(0),
            from: Some(env.eoa.address()),
            capabilities: PrepareCallsCapabilities {
                authorize_keys: Default::default(),
                meta: Meta { fee_token: env.erc20, fee_payer: None, nonce: Some(U256::ZERO) },
                pre_calls: Default::default(),
                pre_call: Default::default(),
                required_funds: vec![RequiredAsset::new(env.erc20, balance)],
                revoke_keys: Default::default(),
            },
            balance_overrides: Default::default(),
            state_overrides: Default::default(),
            key: Some(key.to_call_key()),
        })
        .await?;

    let signature = key.sign_payload_hash(response.digest).await?;

    let response = env
        .relay_endpoint
        .send_prepared_calls(SendPreparedCallsParameters {
            capabilities: Default::default(),
            context: response.context,
            key: key.to_call_key(),
            signature,
        })
        .await?;

    insta::assert_json_snapshot!(response, {
        ".id" => reduction_from_str::<BundleId>("id"),
    });

    Ok(())
}

#[tokio::test]
async fn test_upgrade_account() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let admin_key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    let response = env
        .relay_endpoint
        .prepare_upgrade_account(PrepareUpgradeAccountParameters {
            address: env.eoa.address(),
            delegation: env.delegation,
            chain_id: None,
            capabilities: UpgradeAccountCapabilities {
                authorize_keys: vec![admin_key.to_authorized()],
            },
        })
        .await?;

    // Sign Intent digest
    let precall_signature = env.eoa.sign_hash(&response.digests.exec).await?;

    // Sign 7702 delegation
    let nonce = env.provider().get_transaction_count(env.eoa.address()).await?;
    let authorization = AuthKind::Auth.sign(&env, nonce).await?;

    // Upgrade account.
    #[allow(clippy::let_unit_value)]
    let response = env
        .relay_endpoint
        .upgrade_account(UpgradeAccountParameters {
            context: response.context,
            signatures: UpgradeAccountSignatures {
                auth: authorization.signature()?,
                exec: precall_signature,
            },
        })
        .await?;

    insta::assert_json_snapshot!(response);

    Ok(())
}

#[tokio::test]
async fn test_get_authorization() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let admin_key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    let response = env
        .relay_endpoint
        .prepare_upgrade_account(PrepareUpgradeAccountParameters {
            address: env.eoa.address(),
            delegation: env.delegation,
            chain_id: None,
            capabilities: UpgradeAccountCapabilities {
                authorize_keys: vec![admin_key.to_authorized()],
            },
        })
        .await?;

    // Sign Intent digest
    let precall_signature = env.eoa.sign_hash(&response.digests.exec).await?;

    // Sign 7702 delegation
    let nonce = env.provider().get_transaction_count(env.eoa.address()).await?;
    let authorization = AuthKind::Auth.sign(&env, nonce).await?;

    // Upgrade account.
    env.relay_endpoint
        .upgrade_account(UpgradeAccountParameters {
            context: response.context,
            signatures: UpgradeAccountSignatures {
                auth: authorization.signature()?,
                exec: precall_signature,
            },
        })
        .await?;

    // Get authorization
    let response = env
        .relay_endpoint
        .get_authorization(GetAuthorizationParameters { address: env.eoa.address() })
        .await?;

    insta::assert_json_snapshot!(response);

    Ok(())
}

#[tokio::test]
async fn test_get_calls_status() -> eyre::Result<()> {
    let config =
        EnvironmentConfig { num_chains: 2, fee_recipient: Address::ZERO, ..Default::default() };
    let env = Environment::setup_with_config(config.clone()).await?;

    // Create a key for signing
    let key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();

    // Account upgrade deployed onchain.
    upgrade_account_lazily(&env, &[key.to_authorized()], AuthKind::Auth).await?;

    let balance =
        IERC20::new(env.erc20, env.provider_for(1)).balanceOf(env.eoa.address()).call().await?
            / uint!(2_U256);
    let response = env
        .relay_endpoint
        .prepare_calls(PrepareCallsParameters {
            calls: vec![Call::transfer(env.erc20, Address::ZERO, uint!(1_U256))],
            chain_id: env.chain_id_for(0),
            from: Some(env.eoa.address()),
            capabilities: PrepareCallsCapabilities {
                authorize_keys: Default::default(),
                meta: Meta { fee_token: env.erc20, fee_payer: None, nonce: Some(U256::ZERO) },
                pre_calls: Default::default(),
                pre_call: Default::default(),
                required_funds: vec![RequiredAsset::new(env.erc20, balance)],
                revoke_keys: Default::default(),
            },
            balance_overrides: Default::default(),
            state_overrides: Default::default(),
            key: Some(key.to_call_key()),
        })
        .await?;

    let signature = key.sign_payload_hash(response.digest).await?;

    let response = env
        .relay_endpoint
        .send_prepared_calls(SendPreparedCallsParameters {
            capabilities: Default::default(),
            context: response.context,
            key: key.to_call_key(),
            signature,
        })
        .await?;

    await_calls_status(&env, response.id).await?;

    let response = env.relay_endpoint.get_calls_status(response.id).await?;

    // TODO: these redactions are due to the same issue as in `test_prepare_calls`
    insta::assert_json_snapshot!(response, {
        ".id" => insta::dynamic_redaction(move |value, _path| {
            assert_eq!(BundleId::from_str(value.as_str().unwrap()), Ok(response.id));
            "[id]"
        }),
        ".receipts[].logs[].topics[]" => reduction_from_str::<B256>("topic"),
        ".receipts[].logs[].data" => reduction_from_str::<Bytes>("data"),
        ".receipts[].logs[].blockHash" => reduction_from_str::<B256>("blockHash"),
        ".receipts[].logs[].blockNumber" => reduction_from_str::<U64>("blockNumber"),
        ".receipts[].logs[].blockTimestamp" => reduction_from_str::<U64>("blockTimestamp"),
        ".receipts[].logs[].transactionHash" => reduction_from_str::<B256>("transactionHash"),
        ".receipts[].blockHash" => reduction_from_str::<B256>("blockHash"),
        ".receipts[].blockNumber" => reduction_from_str::<U64>("blockNumber"),
        ".receipts[].gasUsed" => reduction_from_str::<U64>("gasUsed"),
        ".receipts[].transactionHash" => reduction_from_str::<B256>("transactionHash"),
    });

    Ok(())
}

#[tokio::test]
async fn test_verify_signature() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let key = KeyWith712Signer::mock_admin_with_key(KeyType::Secp256k1, ADMIN_KEY)?.unwrap();
    upgrade_account_lazily(&env, &[key.to_authorized()], AuthKind::Auth).await?;

    let digest = B256::ZERO;
    let signature = key.sign_payload_hash(digest).await?;

    let response = env
        .relay_endpoint
        .verify_signature(VerifySignatureParameters {
            address: env.eoa.address(),
            chain_id: env.chain_id(),
            digest,
            signature: signature.clone(),
        })
        .await?;

    insta::assert_json_snapshot!(response, {
        ".proof.initPreCall.nonce" => reduction_from_str::<U256>("nonce"),
        ".proof.initPreCall.signature" => reduction_from_str::<Bytes>("signature"),
    });

    Ok(())
}

#[tokio::test]
async fn test_add_faucet_funds() -> eyre::Result<()> {
    let env = Environment::setup().await?;

    let response = env
        .relay_endpoint
        .add_faucet_funds(AddFaucetFundsParameters {
            token_address: env.fee_token,
            address: env.eoa.address(),
            chain_id: env.chain_id(),
            value: U256::ONE,
        })
        .await?;

    insta::assert_json_snapshot!(response, {
        ".transactionHash" => reduction_from_str::<B256>("transactionHash")
    });

    Ok(())
}

/// Creates a reduction that asserts value can be parsed using [`FromStr`], and replaces
/// the value with the provided name.
fn reduction_from_str<T: FromStr>(name: &str) -> insta::internals::Redaction {
    let name = name.to_string();
    insta::dynamic_redaction(move |value, _path| {
        assert!(T::from_str(value.as_str().unwrap()).is_ok());
        format!("[{}]", name)
    })
}

fn sort_json_keys(v: &mut serde_json::Value) {
    use serde_json::Value;
    match v {
        Value::Object(map) => {
            map.sort_keys();
        }
        Value::Array(arr) => {
            for v in arr.iter_mut() {
                sort_json_keys(v);
            }
        }
        _ => {}
    }
}
