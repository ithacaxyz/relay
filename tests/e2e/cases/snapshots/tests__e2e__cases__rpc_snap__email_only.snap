---
source: tests/e2e/cases/rpc_snap.rs
expression: response_email_only
---
{
  "email": "[email_timestamp]"
}
